src/plugins/stake-plugins/generic/pluginInfo/tarotUtils.ts

```diff
@@ -0,0 +1,317 @@
+      const balance = await poolToken.balanceOf(walletAddress)
+      let decimals = 18
+      if (poolTokenType === TarotPoolTokenType.BorrowableA) {
+        decimals = config.token0.decimals
+      } else if (poolTokenType === TarotPoolTokenType.BorrowableB) {
+        decimals = config.token1.decimals
+      }
+      return {
+        amount: balance.mul(exchangeRate).div(TEN_18),
+        decimals: BigNumber.from(decimals)
+      }
+    },
+
+    async getBorrowed(poolTokenType: TarotPoolTokenType): Promise<BigAmount> {
+      const contracts = await this.getContracts(poolTokenType)
+      const borrowable = contracts[0] as TarotBorrowable
+      const [balance, accrualTimestamp, borrowRate] = await Promise.all([
+        borrowable.borrowBalance(walletAddress),
+        borrowable.accrualTimestamp(),
+        borrowable.borrowRate()
+      ])
+
+      let decimals = 18
+      if (poolTokenType === TarotPoolTokenType.BorrowableA) {
+        decimals = config.token0.decimals
+      } else if (poolTokenType === TarotPoolTokenType.BorrowableB) {
+        decimals = config.token1.decimals
+      }
+      return {
+        amount: balance.add(
+          balance
+            .mul(60000 + Date.now() - accrualTimestamp * 1000)
+            .mul(borrowRate)
+            .div(TEN_18)
+            .div(1000)
+        ),
+        decimals: BigNumber.from(decimals)
+      }
+    },
+
+    parseNumber(bigAmount: BigAmount | undefined): number {
+      if (!bigAmount) {
+        return 0
+      }
+      return parseFloat(formatUnits(bigAmount.amount, bigAmount.decimals))
+    },
+
+    async getValuesFromPrice(
+      changes: typeof this.NO_CHANGES,
+      priceA: number,
+      priceB: number
+    ): Promise<{ valueCollateral: number; valueA: number; valueB: number }> {
+      const [valueCollateralPart, amountAPart, amountBPart] = await Promise.all([
+        this.getDeposited(TarotPoolTokenType.Collateral),
+        this.getBorrowed(TarotPoolTokenType.BorrowableA),
+        this.getBorrowed(TarotPoolTokenType.BorrowableB)
+      ])
+      const valueCollateral = this.parseNumber(valueCollateralPart) + changes.changeCollateral
+      const amountA = this.parseNumber(amountAPart) + changes.changeBorrowedA
+      const amountB = this.parseNumber(amountBPart) + changes.changeBorrowedB
+      const valueA = amountA * priceA
+      const valueB = amountB * priceB
+      return {
+        valueCollateral: valueCollateral > 0 ? valueCollateral : 0,
+        valueA: valueA > 0 ? valueA : 0,
+        valueB: valueB > 0 ? valueB : 0
+      }
+    },
+
+    async getMarketValues(changes: typeof this.NO_CHANGES): Promise<{ valueCollateral: number; valueA: number; valueB: number }> {
+      const [priceA, priceB] = await this.getMarketPriceDenomLP()
+      return await this.getValuesFromPrice(this.NO_CHANGES, priceA, priceB)
+    },
+
+    async getMaxDeleverage(): Promise<number> {
```

Comment:
This function is dense and could use some comment TLC

Reply:
it's copied from tarot.to